Google.Api.Generator/Generation/MethodDetails.cs

```diff
@@ -476,28 +534,73 @@ private IEnumerable<RoutingHeader> ReadRoutingHeaders(HttpRule http, MessageDesc
                 {
                     foreach (var path in ExtractBracedPaths(url))
                     {
-                        var desc = requestDesc;
-                        var fields = new List<FieldDescriptor>();
-                        FieldDescriptor finalField = null;
-                        foreach (var fieldName in path.Split('.'))
-                        {
-                            var field = finalField = desc.FindFieldByName(fieldName);
-                            desc = field.FieldType == FieldType.Message ? field.MessageType : null;
-                            if (field == null)
-                            {
-                                throw new InvalidOperationException($"Invalid path in http url: '{path}'. '{fieldName}' does not exist.");
-                            }
-                            fields.Add(field);
-                        }
-                        if (finalField.FieldType != FieldType.String)
-                        {
-                            throw new InvalidOperationException($"Path in http url must resolve to a string field: '{path}'.");
-                        }
-                        yield return new RoutingHeader(WebUtility.UrlEncode(path), fields);
+                        yield return RoutingHeader.CreateOneFieldNoRegex(WebUtility.UrlEncode(path),
+                            SplitVerifyFieldPath(path, requestDesc));
                     }
                 }
             }
 
+            ExplicitRoutingHeaderPrecursor ExtractHeaderPrecursor(RoutingParameter param, int order)
+            {
+                if (param.PathTemplate == string.Empty)
+                {
+                    return new ExplicitRoutingHeaderPrecursor
+                    {
+                        FieldPath = param.Field,
+                        RegexString = $"^{ResourcePattern.DoubleWildcardStandaloneRegexStr}$",
+                        HeaderName = param.Field,
+                        NoRegexMatchingNeeded = true 
+                    };
+                }
+
+                var pattern = new ResourcePattern(param.PathTemplate);
+
+                var parameterName = pattern.ParameterNames.SingleOrDefault();
+                if (parameterName == null)
+                {
+                    var msg = $"The explicit routing header pattern {param.PathTemplate} should have exactly one non-multivariate ResourceId segment";
+                    throw new InvalidOperationException(
+                        msg);
+                }
+
+                // The full pattern should match the entire field and allow for an
+                // optional `/` at the end.
+                var patternRegex = pattern.FullFieldRegexString;
```

Comment:
I think it would be fine to add that as a comment just above the specific property setter. Not a blocker, but something to consider next time we're in this file.